src/handlers.js

```diff
@@ -1,104 +1,90 @@
-                        .filter(([pn, arg]) => !arg && requiredParams.includes(pn))
-                        .map(([param]) => `queryparam ${param} must have a value supplied`),
-                ];
+    const queries = report.queries
+        .map((q) => db.query(q.query, queryArgs)
+            .then((result) => ({ [q.name]: result })));
+    const result = await Promise.all(queries);
+    render(Object.assign({}, ...result));
+};
 
-                ctx.assert(
-                    requestErrors.length === 0,
-                    400,
-                    {
-                        message: 'Errors in request',
-                        errors: requestErrors,
-                    },
-                );
+const createReportHandler = async (db, report) => {
+    const render = ejs.compile(report.template);
+    await validateReport(render, report, db);
 
-                const queryArgs = {
-                    ...optionalParamDefaults,
-                    // Filter out empty strings as these are optional parameters with no supplied
-                    // value. We'll treat these as null, rather than an empty string.
-                    ...Object.fromEntries(Array.from(ctx.request.URL.searchParams.entries()).filter(([, v]) => v !== '')),
-                };
+    return {
+        get: async (ctx) => {
+            const errors = [];
+            const queryArgs = {};
+            for (const param of (report.endpoint.params || [])) {
+                if (!ctx.request.query[param.name] && param.required) {
+                    errors.push(`Missing parameter in querystring: ${param.name}`);
+                } else {
+                    queryArgs[param.name] = ctx.request.query[param.name] || param.default;
+                }
+            }
 
-                const queries = Object.entries(template.dataSource.data)
-                    .map(([k, v]) => ctx.db.query(v, queryArgs)
-                        .then((result) => ({ [k]: result })));
-                const result = await Promise.all(queries);
-                ctx.state.html = template.render(Object.assign({}, ...result));
-            },
-        };
-        return [`/${route}`, handler];
-    };
+            ctx.assert(
+                errors.length === 0,
+                400,
+                {
+                    message: 'Errors in request',
+                    errors,
+                },
+            );
 
-    return Object.fromEntries(Object.entries(reportTemplates).map(createReportHandler));
+            const queries = report.queries
+                .map((q) => ctx.db.query(q.query, queryArgs)
+                    .then((result) => ({ [q.name]: result })));
+            const result = await Promise.all(queries);
+            try {
+                const html = render(Object.assign({}, ...result));
+                await formatResponse(ctx, html);
+            } catch (e) {
+                ctx.state.logger.log(e);
```

Comment:
Will this return an error to the caller? Should it?

Reply:
We shouldn't pass any errors from this method to a user, as we already validate the report custom resource upon creation.
So any other errors are internal ones.

Reply:
It looks to me that if `formatResponse` throws, the exception will be caught here, and the caller will simply be left waiting for an HTTP response which it will never receive. Sorry if I've misunderstood and I'm causing toil.

Reply:
It won't hang the execution of the request. Also I've just added an HTTP 500 status code.

Reply:
Yes you're right. And yeah that was the right thing to do. Just needed to give the client a bit of information about what they were getting.